js.go

```diff
@@ -422,11 +433,23 @@ func (js *js) PublishMsg(m *Msg, opts ...PubOpt) (*PubAck, error) {
 	}
 
 	if err != nil {
-		if err == ErrNoResponders {
-			err = ErrNoStreamResponse
+		for r := 0; err == ErrNoResponders && r < o.rnum; r++ {
+			// To protect against small blips in leadership changes etc, if we get a no responders here retry.
+			time.Sleep(o.rwait)
+			if o.ttl > 0 {
+				resp, err = js.nc.RequestMsg(m, time.Duration(o.ttl))
```

Comment:
Note that the call below uses the context that is created at the beginning, so it is likely that the deadline is the original one, whereas here you make each request with the original timeout, so maybe that should be reduced by the time already spent? If not, there will be some difference in behavior between context/timeout.

Reply:
I think this might be fine, I like how the context controls the max duration of the retries logic. For example:

```go
ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
defer cancel()
_, err := js.Publish(subj, msg, nats.Context(ctx), nats.RetryWait(500*time.Millisecond), nats.RetryAttempts(30))
```

Above would mean to attempt to publish for 10 seconds max, although the retries would potentially take 15s but the context still sets the deadline.  Also means that possible to have infinite retries, so would retry as needed until the context expires:

```go
ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
defer cancel()
_, err := js.Publish(subj, msg, nats.Context(ctx), nats.RetryWait(500*time.Millisecond), nats.RetryAttempts(-1))
```

Reply:
I will do a rough estimation here subtracting out o.rwait and bailing if local ttl in the loop goes negative. Will post separate CL for ease of review but can squash before merging.